### ISSUE TITLE
Research

### ISSUE THREAD

**Issue by Omkarai2002** [Omkarai2002/Argos-Agentic#1]
1. Discussion on how the acknowledgement is going to be passed .
2. Web socket integration .
3. understanding the input as well as output parameter requirements.
4. tech stack to be used 
5. models to be used 
6.Low level Research on everything in detail to be used 
7. databases to be hit for AI side as well as software side 
8. features to be added for current and future phase
9. finalize the data intake format for the ngram as well as the SNN model.
10. Low level architecture to be build.

**Comment by Omkarai2002** [comment]
1. System Architecture: The Two-Layer Model
To achieve high relevance without heavy computational overhead, the system splits intelligence into static and dynamic layers.
Layer 1: Base Intelligence (Static)
Components: N-gram models and Small Neural Networks (SNN).
Function: Provides the foundational linguistic understanding. This layer is pre-trained and does not change during client-side use.
Layer 2: Personal Memory (Dynamic)
Components: IndexedDB (Browser-side storage).
Function: Captures per-user history and per-website behavior.
Key Features:
Fast Look-up: Optimized for real-time retrieval.
No Retraining: Updates dynamically by indexing user input without needing back-end model fine-tuning.
Ranking Only: This layer is primarily used to re-rank suggestions provided by Layer 1.
The Scoring Formula
The final suggestion ranking is determined by a weighted sum:
Final\_Score = Model\_Score + Ngram\_Score + Personal\_History\_Boost
2. Core Services
The system provides a suite of features via a WebSocket connection between the Browser/App and the Suggestion Service.
Word Completion: Utilizing a Trie data structure for efficient prefix matching and the Python wordfreq module for frequency-based completion.
Auto-Correction: Continuous spell-check monitoring and correction indications.
Predictive Text: Powered by the N-gram and SNN models mentioned in Layer 1.
3. Data Engineering & Training
"Good data > Fancy models." The effectiveness of the N-gram and SNN models relies on rigorous preprocessing.
Prerequisites for Training (training.txt)
Lowercasing: Convert all text to lowercase for uniformity.
Cleaning: Remove junk symbols and non-alphanumeric characters.
Normalization: Standardize spellings (e.g., "u" to "you").
Deduplication: Remove empty lines and redundant whitespace.
Volume: Requires a minimum of 2,000 to 5,000 sentences for a baseline.
The Golden Rules of Optimization
Rule 1: If the model fails due to a lack of exposure to specific contexts \rightarrow Add more data.
Rule 2: If the model fails even on common, high-frequency examples \rightarrow Tune the model architecture/parameters.
4. Performance Optimization
Debouncing (Websockets)
To prevent network congestion and reduce server load by up to 90%, the system implements debouncing:
Time-based: Send events only every 50–100ms.
Trigger-based: Send requests only on spacebar presses or upon specific word completion triggers.

**Comment by Omkarai2002** [comment]
<img width="903" height="652" alt="Image" src="https://github.com/user-attachments/assets/45920dc0-567c-4093-83cc-0483ad2758b2" />

JSON Field Validation and Completion Workflow
Overview

This workflow is responsible for validating mission-specific parameters inside the generated JSON. It ensures that all required operational fields are present, within acceptable thresholds, and explicitly confirmed before the mission data is considered complete. If any required field is missing, the system enters a controlled clarification loop instead of guessing values.

Step-by-Step Workflow Explanation
1. JSON Output Entry

The workflow begins with a structured JSON object generated by the system. At this stage, the JSON is already syntactically valid but may still be partial or incomplete in terms of required mission parameters.

2. Completeness Check (Filled vs Partial)

The system evaluates whether all required fields for the selected mission type are present.

Filled (YES):
All mandatory fields exist in the JSON.

Partial (NO):
One or more required fields are missing or empty.

This decision determines the next validation path.

Filled Path (All Required Fields Present)
3. Parameter Threshold Validation

If the JSON is marked as filled, the system performs field-level threshold checks to ensure operational safety and correctness.

The following checks are applied sequentially:

Check Altitude
Ensures the altitude is within defined minimum and maximum operational limits.

Check Radius
Verifies that the radius value is within acceptable bounds for the mission type.

Check Speed
Confirms that the speed parameter is safe and within allowed limits.

Check Battery (Planned for Next Phase)
Battery feasibility checks are noted as a future enhancement and are not enforced at this stage.

These checks ensure that even complete data is safe and usable, not just present.

4. Validated JSON Output

Once all threshold checks pass, the JSON is considered operationally valid and is passed forward as a validated JSON output.

5. Coordinate Extraction and Validation

The validated JSON is then sent to a coordinate extraction function, which isolates latitude and longitude values. These coordinates can subsequently be validated against external systems or geofence data retrieved from persistent storage (e.g., RDS).

This step ensures geographic correctness and traceability.

Partial Path (Missing Required Fields)
6. Identify Missing Fields

If the JSON is marked as partial, the system identifies exactly which fields are empty or missing.

This step is deterministic and schema-driven, ensuring no ambiguity in what information is required.

7. Frame Static Clarification Questions

Based on the missing fields, the system generates static, targeted questions for the user. These questions are predefined and specific, avoiding open-ended or ambiguous prompts.

Example intent:

Ask only for missing altitude

Ask only for missing speed

Ask only for missing radius

8. User Input Collection

The user provides responses to the clarification questions. These inputs are captured and merged back into the existing JSON structure.

The system does not overwrite existing valid fields.

9. Loop Back to Validation

After updating the JSON with user-provided values, the workflow loops back to the Filled vs Partial check, re-evaluating completeness and thresholds.

This loop continues until:

All required fields are present and valid, or
